panic on building through ssh dialer

```
» /tmp/docker -H ssh://docker@35.161.xx.xx build .                                                                               
Enter passphrase for key '/Users/tonistiigi/.ssh/id_rsa':
[+] Building 2.1s (0/2)
[+] Building 3.0s (6/6) FINISHED
 => [internal] load Dockerfile                                                                                                       2.4s
 => => transferring dockerfile: 34B                                                                                                  0.1s
 => [internal] load .dockerignore                                                                                                    2.4s
 => => transferring context: 2B                                                                                                      0.1s
 => [internal] load metadata for docker.io/library/alpine:latest                                                                     0.6s
 => [1/2] FROM docker.io/library/alpine@sha256:7043076348bf5040220df6ad703798fd8593a0918d06d3ce30c6c93be117e430                      0.0s
 => CACHED [2/2] RUN ls -l                                                                                                           0.0s
 => exporting to image                                                                                                               0.0s
 => => exporting layers                                                                                                              0.0s
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x44988ee]

goroutine 51 [running]:
os.(*ProcessState).os.exited(...)
	/usr/local/go/src/os/exec_posix.go:70
os.(*ProcessState).Exited(...)
	/usr/local/go/src/os/exec.go:144
github.com/docker/cli/cli/connhelper.(*commandConn).killIfStdioClosed(0xc4205a69c0, 0x0, 0x0)
	/Users/tonistiigi/gocode/src/github.com/docker/cli/cli/connhelper/connhelper.go:112 +0xae
github.com/docker/cli/cli/connhelper.(*commandConn).CloseWrite(0xc4205a69c0, 0x0, 0x0)
	/Users/tonistiigi/gocode/src/github.com/docker/cli/cli/connhelper/connhelper.go:170 +0x89
github.com/docker/cli/cli/connhelper.(*commandConn).Close(0xc4205a69c0, 0xc4204eaaa8, 0xc4205b7790)
	/Users/tonistiigi/gocode/src/github.com/docker/cli/cli/connhelper/connhelper.go:186 +0x99
github.com/docker/cli/vendor/golang.org/x/net/http2.(*serverConn).serve(0xc4205f6380)
	/Users/tonistiigi/gocode/src/github.com/docker/cli/vendor/golang.org/x/net/http2/server.go:824 +0xb43
github.com/docker/cli/vendor/golang.org/x/net/http2.(*Server).ServeConn(0xc4200c60c0, 0x5887b00, 0xc42000e008, 0xc420079d68)
	/Users/tonistiigi/gocode/src/github.com/docker/cli/vendor/golang.org/x/net/http2/server.go:435 +0x75f
github.com/docker/cli/vendor/github.com/moby/buildkit/session.serve(0x5876b00, 0xc420321240, 0xc4205f6000, 0x5887b00, 0xc42000e008)
	/Users/tonistiigi/gocode/src/github.com/docker/cli/vendor/github.com/moby/buildkit/session/grpc.go:24 +0xf8
github.com/docker/cli/vendor/github.com/moby/buildkit/session.(*Session).Run(0xc4203e2070, 0x5876b00, 0xc420321240, 0xc420079f60, 0x0, 0x0)
	/Users/tonistiigi/gocode/src/github.com/docker/cli/vendor/github.com/moby/buildkit/session/session.go:103 +0x6f7
github.com/docker/cli/cli/command/image.runBuildBuildKit.func1(0x8, 0x577ae98)
	/Users/tonistiigi/gocode/src/github.com/docker/cli/cli/command/image/build_buildkit.go:145 +0xa1
github.com/docker/cli/vendor/golang.org/x/sync/errgroup.(*Group).Go.func1(0xc420046240, 0xc4205ea0a0)
	/Users/tonistiigi/gocode/src/github.com/docker/cli/vendor/golang.org/x/sync/errgroup/errgroup.go:58 +0x57
created by github.com/docker/cli/vendor/golang.org/x/sync/errgroup.(*Group).Go
	/Users/tonistiigi/gocode/src/github.com/docker/cli/vendor/golang.org/x/sync/errgroup/errgroup.go:55 +0x66
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x44988ee]

goroutine 66 [running]:
os.(*ProcessState).os.exited(...)
	/usr/local/go/src/os/exec_posix.go:70
os.(*ProcessState).Exited(...)
	/usr/local/go/src/os/exec.go:144
github.com/docker/cli/cli/connhelper.(*commandConn).killIfStdioClosed(0xc4205a69c0, 0xc4205d2570, 0xc4205d2501)
	/Users/tonistiigi/gocode/src/github.com/docker/cli/cli/connhelper/connhelper.go:112 +0xae
github.com/docker/cli/cli/connhelper.(*commandConn).CloseRead(0xc4205a69c0, 0x0, 0xc42009db00)
	/Users/tonistiigi/gocode/src/github.com/docker/cli/cli/connhelper/connhelper.go:151 +0x89
github.com/docker/cli/cli/connhelper.(*commandConn).Close(0xc4205a69c0, 0x48d9601, 0xc420320001)
	/Users/tonistiigi/gocode/src/github.com/docker/cli/cli/connhelper/connhelper.go:183 +0x2f
github.com/docker/cli/vendor/github.com/moby/buildkit/session.serve.func1(0x5876b00, 0xc420321240, 0x5887b00, 0xc42000e008)
	/Users/tonistiigi/gocode/src/github.com/docker/cli/vendor/github.com/moby/buildkit/session/grpc.go:21 +0x5c
created by github.com/docker/cli/vendor/github.com/moby/buildkit/session.serve
	/Users/tonistiigi/gocode/src/github.com/docker/cli/vendor/github.com/moby/buildkit/session/grpc.go:19 +0x61

```

3ea56aa0cace40267402393d245eaddf30de06cb

@AkihiroSuda

--- a/cli/connhelper/connhelper.go
+++ b/cli/connhelper/connhelper.go
@@ -82,6 +82,7 @@ func newCommandConn(ctx context.Context, cmd string, args ...string) (net.Conn, 
 // commandConn implements net.Conn
 type commandConn struct {
 	cmd           *exec.Cmd
+	cmdMutex      sync.Mutex
 	stdin         io.WriteCloser
 	stdout        io.ReadCloser
 	stderrMu      sync.Mutex
@@ -102,6 +103,7 @@ func (c *commandConn) killIfStdioClosed() error {
 		return nil
 	}
 	var err error
+	c.cmdMutex.Lock()
 	// NOTE: maybe already killed here
 	if err = c.cmd.Process.Kill(); err == nil {
 		err = c.cmd.Wait()
@@ -113,11 +115,14 @@ func (c *commandConn) killIfStdioClosed() error {
 			err = nil
 		}
 	}
+	c.cmdMutex.Unlock()
 	return err
 }
 
 func (c *commandConn) onEOF(eof error) error {
+	c.cmdMutex.Lock()
 	werr := c.cmd.Wait()
+	c.cmdMutex.Unlock()
 	if werr == nil {
 		return eof
 	}
@@ -131,6 +136,7 @@ func ignorableCloseError(err error) bool {
 	errS := err.Error()
 	ss := []string{
 		os.ErrClosed.Error(),
+		"process already finished",
 	}
 	for _, s := range ss {
 		if strings.Contains(errS, s) {
@@ -148,7 +154,10 @@ func (c *commandConn) CloseRead() error {
 	c.stdioClosedMu.Lock()
 	c.stdoutClosed = true
 	c.stdioClosedMu.Unlock()
-	return c.killIfStdioClosed()
+	if err := c.killIfStdioClosed(); err != nil && !ignorableCloseError(err) {
+		logrus.Warnf("commandConn.CloseRead: %v", err)
+	}
+	return nil
 }
 
 func (c *commandConn) Read(p []byte) (int, error) {
@@ -167,7 +176,10 @@ func (c *commandConn) CloseWrite() error {
 	c.stdioClosedMu.Lock()
 	c.stdinClosed = true
 	c.stdioClosedMu.Unlock()
-	return c.killIfStdioClosed()
+	if err := c.killIfStdioClosed(); err != nil && !ignorableCloseError(err) {
+		logrus.Warnf("commandConn.CloseWrite: %v", err)
+	}
+	return nil
 }
 
 func (c *commandConn) Write(p []byte) (int, error) {

